src/internet/model/tcp-socket-base.cc

```diff
@@ -4156,4 +4098,592 @@ RttHistory::RttHistory (const RttHistory& h)
+    {
+      // Assuming packet size is constant
+      if (actualSeq < m_highestSeq)
+        {
+          // Loss in retransmit
+          // We don't know the byte ordering from received buffer, so this is just an estimate
+          // Assuming the loss rate during retransmit would be low
+          m_lossCounts.push_back (std::make_pair (Simulator::Now ().GetMilliSeconds (), 1));
+
+        }
+      else if (actualSeq > m_highestSeq + p->GetSize ())
+        {
+          // Loss in new packets
+          uint32_t loss_packets = (actualSeq - m_highestSeq - p->GetSize ()) / p->GetSize ();
+          m_lossCounts.push_back (std::make_pair (Simulator::Now ().GetMilliSeconds (), loss_packets));
+        }
+    }
+
+  // Update m_lastReceivedSeq to record the last sequence number received
+  m_lastReceivedSeq = actualSeq;
+
+  // Update the highest sequence number
+  if (m_lastReceivedSeq > m_highestSeq)
+    {
+      m_highestSeq = m_lastReceivedSeq;
+    }
+}
+
+double
+TcpSocketBase::EstimateFairShareCalc (Ptr<Packet> p, SequenceNumber32 seq,
+                                      Time rtt)
+{
+  NS_LOG_FUNCTION (this << p << seq << rtt);
+  EstimateLoss (p, seq);
+
+  // Since BBR's probeBW lasts for eight RTT cycles, here we also measure the
+  // throughput for last 8 * RTT
+  uint32_t num_rtts = 8;
+  double rttS = rtt.GetSeconds ();
+  // First remove packets and losses older than 8 * RTT
+  int64_t cutoff = Simulator::Now ().GetMilliSeconds () - (num_rtts * rttS);
+
+  while (m_arrivalTimes.size () > 0 &&
+         m_arrivalTimes.front ().GetMilliSeconds () < cutoff)
+    {
+      m_arrivalTimes.pop_front ();
+    }
+
+  while (m_lossCounts.size () > 0 &&
+         std::get<0> (m_lossCounts.front ()) < cutoff)
+    {
+      m_lossCounts.pop_front ();
+    }
+
+  // Check if BBR enters probeRTT phase - it compares the number of packets
+  // received in last RTT against the average of 8 RTTs.
+  int64_t last_rtt = Simulator::Now ().GetMilliSeconds () - rttS;
+  uint32_t last_rtt_packets = 0;
+  for (auto tput_iter = m_arrivalTimes.rbegin ();
+       tput_iter != m_arrivalTimes.rend () &&
+         (*tput_iter).GetMilliSeconds () > last_rtt;
+       ++tput_iter)
+    {
+      ++last_rtt_packets;
+    }
+
+  // If last RTT received one tenth of its average throughput, reply ACK as soon
+  // as possible.
+  if (last_rtt_packets * num_rtts < (m_arrivalTimes.size () / 10))
+    {
+      return -1;
+    }
+
+  // Count the number of packet losses.
+  int loss_count = 0;
+  for (const auto& loss : m_lossCounts)
+    {
+      loss_count += std::get<1> (loss);
+    }
+
+  // Constant C.
+  double c = 1.2247;
+  double loss_rate = loss_count / (1.0 * (loss_count + m_arrivalTimes.size ()));
+  double fairTput = 0;
+  switch (m_fairShareType)
+    {
+    case FairShareEstimationType::Mathis:
+      if (loss_rate > 0)
+        {
+          fairTput = c / (rttS * sqrt (loss_rate)); // In number of packets per second
```

Comment:
I just realized that this would overestimate the fair throughput. Imagine the loss count of last eight RTTS (latest to earliest) to be (10, 10, 10, 0, 0, 0, 0, 0), the moving average after 6 RTT would be (0, 0, 0, 0, 0, 0, 10, 10). The estimation is even higher after the next RTT. I tested it and removing the zero-loss RTTs at the end would give a more accurate fair share estimate.

I will update this once your code is merged in.